pkg/services/ngalert/api/authorization.go

```diff
@@ -162,3 +171,90 @@ func (api *API) authorize(method, path string) web.Handler {
 
 	panic(fmt.Sprintf("no authorization handler for method [%s] of endpoint [%s]", method, path))
 }
+
+// GetDatasourceScopesFromAlertRule extracts data source scopes from an alert rule
+func getEvaluatorForAlertRule(rule *ngmodels.AlertRule) ac.Evaluator {
+	scopes := make([]ac.Evaluator, 0, len(rule.Data))
+	for _, query := range rule.Data {
+		if query.QueryType == expr.DatasourceType || query.DatasourceUID == expr.OldDatasourceUID {
+			continue
+		}
+		scopes = append(scopes, ac.EvalPermission(datasources.ActionQuery, dashboards.ScopeFoldersProvider.GetResourceScopeUID(query.DatasourceUID)))
+	}
+	return ac.EvalAll(scopes...)
+}
+
+// authorizeRuleChanges analyzes changes in the rule group, determines what actions the user is trying to perform and check whether those actions are authorized.
+// If the user is not authorized to perform the changes the function returns ErrAuthorization with description of what action is not authorized. If evaluator function returns error, the function returns it
+func authorizeRuleChanges(namespace *models.Folder, changes *changes, evaluator func(evaluator ac.Evaluator) (bool, error)) error {
+	namespaceScope := dashboards.ScopeFoldersProvider.GetResourceScope(strconv.FormatInt(namespace.Id, 10))
+	if len(changes.Delete) > 0 {
+		allowed, err := evaluator(ac.EvalPermission(ac.ActionAlertingRuleDelete, namespaceScope))
+		if err != nil {
+			return fmt.Errorf("failed to perform user authorization: %w", err)
+		}
+		if !allowed {
+			return fmt.Errorf("%w user cannot delete alert rules that belong to folder %s", ErrAuthorization, namespace.Title)
+		}
+	}
+
+	// check that the user can create alerts in
```

Comment:
```suggestion
```

Was this a leftover? I think it's somewhat clear based on the function description, no need for a comment.